File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/core/binding/ArrayBindingAdapter.java
package com.skedgo.tripkit.ui.core.binding;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import com.skedgo.tripkit.common.model.RealtimeAlert;
import com.skedgo.tripkit.ui.BR;
import com.skedgo.tripkit.ui.views.TripSegmentAlertView;

import java.util.ArrayList;
import java.util.List;

public final class ArrayBindingAdapter {
    @BindingAdapter({"entries", "layout"})
    public static <T> void setEntries(ViewGroup viewGroup,
                                      List<T> entries, int layoutId) {
        viewGroup.removeAllViews();
        if (entries != null) {
            LayoutInflater inflater = (LayoutInflater)
                    viewGroup.getContext()
                            .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            for (int i = 0; i < entries.size(); i++) {
                T entry = entries.get(i);
                ViewDataBinding binding = DataBindingUtil
                        .inflate(inflater, layoutId, viewGroup, true);
                binding.setVariable(BR.viewModel, entry);
            }
        }
    }

    @BindingAdapter({"alerts"})
    public static void setAlertEntries(ViewGroup viewGroup, ArrayList<RealtimeAlert> alerts) {
        if (viewGroup instanceof TripSegmentAlertView) {
            TripSegmentAlertView view = (TripSegmentAlertView) viewGroup;
            view.setAlerts(alerts);
        }
    }
}

File: TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/core/binding/ArrayBindingAdapter.kt
package com.skedgo.tripkit.ui.core.binding

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.BindingAdapter
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.skedgo.tripkit.common.model.realtimealert.RealtimeAlert
import com.skedgo.tripkit.ui.BR
import com.skedgo.tripkit.ui.views.TripSegmentAlertView

@BindingAdapter("entries", "layout")
fun <T> setEntries(
    viewGroup: ViewGroup,
    entries: List<T>?,
    layoutId: Int
) {
    viewGroup.removeAllViews()
    if (entries != null) {
        val inflater = viewGroup.context
            .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        for (entry in entries) {
            val binding = DataBindingUtil
                .inflate<ViewDataBinding>(inflater, layoutId, viewGroup, true)
            binding.setVariable(BR.viewModel, entry)
        }
    }
}

@BindingAdapter("alerts")
fun setAlertEntries(viewGroup: ViewGroup, alerts: ArrayList<RealtimeAlert>) {
    if (viewGroup is TripSegmentAlertView) {
        viewGroup.setAlerts(alerts)
    }
}
